.. _pyani-citations:

================
Citing ``pyani``
================

We would be grateful if you could please cite the following manuscript in your work if you have found ``pyani`` useful:

    Pritchard *et al.* (2016) "Genomics and taxonomy in diagnostics for food security: soft-rotting enterobacterial plant pathogens" *Anal. Methods*, 2016, **8**, 12-24 DOI: 10.1039/C5AY02550H

.. code-block:: latex

    @Article{C5AY02550H,
    author ="Pritchard, Leighton and Glover, Rachel H. and Humphris, Sonia and Elphinstone, John G. and Toth, Ian K.",
    title  ="Genomics and taxonomy in diagnostics for food security: soft-rotting enterobacterial plant pathogens",
    journal  ="Anal. Methods",
    year  ="2016",
    volume  ="8",
    issue  ="1",
    pages  ="12-24",
    publisher  ="The Royal Society of Chemistry",
    doi  ="10.1039/C5AY02550H",
    url  ="http://dx.doi.org/10.1039/C5AY02550H",
    abstract  ="Soft rot Enterobacteriaceae (SRE) are bacterial plant pathogens that cause blackleg{,} wilt and soft rot
    diseases on a broad range of important crop and ornamental plants worldwide. These organisms (spanning the genera Erwinia{,}
    Pectobacterium{,} Dickeya{,} and Pantoea) cause significant economic and yield losses in the field{,} and in storage. They
    are transmissible through surface water{,} by trade and other movement of plant material and soil{,} and in some cases are
    subject to international legislative and quarantine restrictions. Effective detection and diagnosis in support of food
    security legislation and epidemiology is dependent on the ability to classify pathogenic isolates precisely. Diagnostics and
    classification are made more difficult by the influence of horizontal gene transfer on phenotype{,} and historically complex
    and sometimes inaccurate nomenclatural and taxonomic assignments that persist in strain collections and online sequence
    databases. Here{,} we briefly discuss the relationship between taxonomy{,} genotype and phenotype in the SRE{,} and their
    implications for diagnostic testing and legislation. We present novel whole-genome classifications of the SRE{,}
    illustrating inconsistencies between the established taxonomies and evidence from completely sequenced isolates. We conclude
    with a perspective on the future impact of widespread whole-genome sequencing and classification methods on detection and
    identification of bacterial plant pathogens in support of legislative and policy efforts in food security."}


.. TIP::
    When citing ``pyani`` in your text, please consider quoting the precise release version (if you are using a specific release), or the specific commit hash (if you are using the 'bleeding-edge' development verison), e.g.

    ``"...using PYANI (v0.3.0)..."``

    or

    ``"...using PYANI (commit bb1cb5c)..."``

    This will enable precise reproduction of this part of your work by others.

-----------------------------
Publications citing ``pyani``
-----------------------------

If you are using ``pyani``, you are in good company. These authors and manuscripts have employed ``pyani`` to help with their whole-genome classification:

(If I've missed out your paper, please do put in a pull request on ``GitHub`` or get in touch and I'll be happy to add it!)

^^^^
2020
^^^^

* Albert *et al.* (2020) "Comparative Pangenomics of the Mammalian Gut Commensal *Bifidobacterium longum*" *Microorganisms* `doi:10.3390/microorganisms8010007 <https://doi.org/10.3390/microorganisms8010007>`_
* Ibarra Caballero *et al.* (2020) "Genome comparison and transcriptome analysis of the invasive brown root rot pathogen, *Phellinus noxius*, from different geographic regions reveals potential enzymes associated with degradation of different wood substrates" *Fungal Biology* `doi:10.1016/j.funbio.2019.12.007 <https://doi.org/10.1016/j.funbio.2019.12.007>`_
* Kuleshov *et al.* (2020) "Whole genome sequencing of *Borrelia miyamotoi* isolate Izh-4: reference for a complex bacterial genome" *BMC Genomics* `doi:10.1186/s12864-019-6388-4 <https://doi.org/10.1186/s12864-019-6388-4>`_
* Lacault *et al.* (2020) "Zucchini vein clearing disease is caused by several lineages within *Pseudomonas syringae* species complex." *Phytopathology* `doi:10.1094/PHYTO-07-19-0266-R <https://doi.org/10.1094/PHYTO-07-19-0266-R>`_
* Wang *et al.* (2020) "Comparative Genomics Analysis of *Lactobacillus ruminis* from Different Niches" *Genes* `doi:10.3390/genes11010070 <https://doi.org/10.3390/genes11010070>`_

^^^^
2019
^^^^

* Accetto & Avgustin (2019) "The diverse and extensive plant polysaccharide degradative apparatuses of the rumen and hindgut *Prevotella* species: A factor in their ubiquity?" *Syst. Appl. Microbiol.* `doi:j.syapm.2018.10.001 <https://doi.org/j.syapm.2018.10.001>`_
* Acevedo *et al.* (2019) "*Bacillus clarus* sp. nov. is a new *Bacillus cereus* group species isolated from soil" *BioRxiv* `doi:10.1101/508077 <https://doi.org/10.1101/508077>`_
* Alberoni *et al.* (2019) "*Bifidobacterium xylocopae* sp. nov. and *Bifidobacterium aemilianum *sp. nov., from the carpenter bee (*Xylocopa violacea*) digestive tract" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2018.11.005 <https://doi.org/10.1016/j.syapm.2018.11.005>`_
* Alex & Antunes (2019) "Whole-Genome Comparisons Among the Genus *Shewanella* Reveal the Enrichment of Genes Encoding Ankyrin-Repeats Containing Proteins in Sponge-Associated Bacteria" *Front. Microbiol.* `doi:10.3389/fmicb.2019.00005 <https://doi.org/10.3389/fmicb.2019.00005>`_
* Alex & Antunes (2019) "Comparative Genomics Reveals Metabolic Specificity of *Endozoicomonas* Isolated from a Marine Sponge and the Genomic Repertoire for Host-Bacteria Symbioses" *Microorganisms* `doi:10.3390/microorganisms7120635 <https://doi.org/10.3390/microorganisms7120635>`_
* Barnier *et al.* (2019) "Description of *Palleronia rufa* sp. nov., a biofilm-forming and AHL-producing *Rhodobacteraceae*, reclassification of *Hwanghaeicola aestuarii* as *Palleronia aestuarii* comb. nov., *Maribius pontilimi* as *Palleronia pontilimi* comb. nov., *Maribius salinus* as *Palleronia salina* comb. nov., *Maribius pelagius* as *Palleronia pelagia* comb. nov. and emended description of the genus *Palleronia*" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2019.126018 <https://doi.org/10.1016/j.syapm.2019.126018>`_
* Bayjanov *et al.* (2019) "Whole genome analysis of *Pandoraea* species strains from cystic fibrosis patients" *Future Microbiology* `doi:10.2217/fmb-2019-0038 <https://doi.org/10.2217/fmb-2019-0038>`_
* Botelho *et al.* (2019) "Combining sequencing approaches to fully resolve a carbapenemase-encoding megaplasmid in a Pseudomonas shirazica clinical strain" *Emerg. Microb. Inf.* `doi:10.1080/22221751.2019.1648182 <https://doi.org/10.1080/22221751.2019.1648182>`_
* Botelho *et al.* (2019) "Megaplasmids on the Rise: Combining Sequencing Approaches to Fully Resolve a Carbapenemase-Encoding Plasmid in a Proposed Novel *Pseudomonas* Species" *BioRxiv* `doi:10.1080/22221751.2019.1648182 <https://doi.org/10.1080/22221751.2019.1648182>`_
* Boukerb *et al.* (2019) "*Campylobacter armoricus* *sp. nov.*, a novel member of the *Campylobacter lari* group isolated from surface water and stools from humans with enteric infection" *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.003836 <https://doi.org/10.1099/ijsem.0.003836>`_
* Briand *et al.* (2019) "A rapid and simple method for assessing and representing genome sequence relatedness" *BioRxiv* `doi:10.1101/569640 <https://doi.org/10.1101/569640>`_
* Cho & Kwak (2019) "Evolution of Antibiotic Synthesis Gene Clusters in the *Streptomyces globisporus* TFH56, Isolated from Tomato Flower" *G3: Genes, Genomes, Genetics* `doi:10.1534/g3.119.400037  <https://dx.doi.org/10.1534/g3.119.400037>`_
* Ciok & Dziewit (2019) "Exploring the genome of Arctic *Psychrobacter* sp. DAB_AL32B and construction of novel *Psychrobacter*-specific cloning vectors of an increased carrying capacity" *Arch. Microbiol.* `doi:10.1007/s00203-018-1595-y <https://doi.org/10.1007/s00203-018-1595-y>`_
* D'Souza *et al.* (2019) "Spatiotemporal dynamics of multidrug resistant bacteria on intensive care unit surfaces" *Nat. Comm.* `doi:10.1038/s41467-019-12563-1 <https://doi.org/10.1038/s41467-019-12563-1>`_
* do Vale *et al.* (2019) "Draft Genome Sequences of Three Novel *Acinetobacter* Isolates from an Irish Commercial Pig Farm" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00919-19 <https://dx.doi.org/10.1128/MRA.00919-19>`_
* Doud *et al.* (2019) "Function-driven single-cell genomics uncovers cellulose-degrading bacteria from the rare biosphere" *ISME J.* `doi:10.1038/s41396-019-0557-y <https://doi.org/10.1038/s41396-019-0557-y>`_
* Du *et al.* (2019) "Characterization of a Linezolid- and Vancomycin-Resistant *Streptococcus suis* Isolate That Harbors optrA and vanG Operons" *Front. Microbiol.* `doi:10.3389/fmicb.2019.02026 <https://doi.org/10.3389/fmicb.2019.02026>`_
* Esposito *et al.* (2019) "Insights into the genome structure of four acetogenic bacteria with specific reference to the Wood–Ljungdahl pathway" *Microbiol. Open* `doi:10.1002/mbo3.938 <https://doi.org/10.1002/mbo3.938>`_
* Falagan *et al.* (2019) "Acidithiobacillus sulfuriphilus sp. nov.: an extremely acidophilic sulfur-oxidizing chemolithotroph isolated from a neutral pH environment" *Int. J. Syst. Evol. Micro.* `doi:0.1099/ijsem.0.003576 <https://doi.org/0.1099/ijsem.0.003576>`_
* Faoro *et al.* (2019) "Genome comparison between clinical and environmental strains of *Herbaspirillum seropedicae* reveals a potential new emerging bacterium adapted to human hosts" *BMC Genomics* `doi:10.1186/s12864-019-5982-9 <https://doi.org/10.1186/s12864-019-5982-9>`_
* Feng *et al.* (2019) "Complete genome sequence of *Hahella* sp. KA22, a prodigiosin-producing algicidal bacterium" *Marine Genomics* `doi:10.1016/j.margen.2019.04.003 <https://doi.org/10.1016/j.margen.2019.04.003>`_
* Gasparrini *et al.* (2019) "Metagenomic signatures of early life hospitalization and antibiotic treatment in the infant gut microbiota and resistome persist long after discharge" *Nature Microbiol.* `doi:10.1038/s41564-019-0550-2 <https://doi.org/10.1038/s41564-019-0550-2>`_
* Ghosh *et al.* (2019) "Reanalysis of *Lactobacillus paracasei* Lbs2 Strain and Large-Scale Comparative Genomics Places Many Strains into Their Correct Taxonomic Position" *Microorganisms* `doi:10.3390/microorganisms7110487 <https://doi.org/10.3390/microorganisms7110487>`_
* Hollensteiner *et al.* (2019) "Complete Genome Sequence of *Marinobacter* sp. Strain JH2, Isolated from Seawater of the Kiel Fjord" *Micro. Res. Ann.* `doi:10.1128/MRA.00596-19 <https://doi.org/10.1128/MRA.00596-19>`_
* Hornung *et al.* (2019) "An in silico survey of *Clostridioides difficile* extrachromosomal elements" *BioRxiv* `doi:10.1101/651539 <https://doi.org/10.1101/651539>`_
* Huang *et al.* (2019) "Genomic differences within the phylum Marinimicrobia: From waters to sediments in the Mariana Trench" *Marine Genomics* `doi:10.1016/j.margen.2019.100699 <https://doi.org/10.1016/j.margen.2019.100699>`_
* Ide *et al.* (2019) "Draft Genome Sequence of *Acidovorax* sp. Strain NB1, Isolated from a Nitrite-Oxidizing Enrichment Culture" *Micro. Res. Ann.* `doi:10.1128/MRA.00547-19 <https://doi.org/10.1128/MRA.00547-19>`_
* Jeong *et al.* (2019) "Chronicle of a Soil Bacterium: *Paenibacillus polymyxa* E681 as a Tiny Guardian of Plant and Human Health" *Front. Microbiol.* `doi:10.3389/fmicb.2019.00467 <https://doi.org/10.3389/fmicb.2019.00467>`_
* Kaminsky *et al.* (2019) "Genomic Analysis of γ-Hexachlorocyclohexane-Degrading *Sphingopyxis lindanitolerans* WS5A3p Strain in the Context of the Pangenome of *Sphingopyxis*" *Genes* `doi:0.3390/genes10090688 <https://doi.org/0.3390/genes10090688>`_
* Khan *et al.* (2019) "Genomic and physiological analyses reveal that extremely thermophilic *Caldicellulosiruptor changbaiensis* deploys unique cellulose attachment mechanisms" *BioRxiv* `doi:10.1101/622977 <https://doi.org/10.1101/622977>`_
* Kirmiz *et al.* (2019) "Comparative genomics guides elucidation of vitamin B12 biosynthesis in novel human associated *Akkermansia*" *BioRxiv* `doi:10.1101/587527 <https://doi.org/10.1101/587527>`_
* Kiu *et al.* (2019) "Genomic analysis on broiler-associated *Clostridium perfringens* strains and exploratory caecal microbiome investigation reveals key factors linked to poultry necrotic enteritis" *Animal Microbiome* `doi:10.1186/s42523-019-0015-1 <https://doi.org/10.1186/s42523-019-0015-1>`_
* Kiu *et al.* (2019) "Phylogenomic analysis of gastroenteritis-associated *Clostridium perfringens* in England and Wales over a 7-year period indicates distribution of clonal toxigenic strains in multiple outbreaks and extensive involvement of enterotoxin-encoding (CPE) plasmids" *Micro. Genom.* `doi:10.1099/mgen.0.000297 <https://doi.org/10.1099/mgen.0.000297>`_
* Lozada *et al.* (2019) "Phage vB_BmeM-Goe8 infecting *Bacillus megaterium* DSM319" *Arch. Virol.* `doi:10.1007/s00705-019-04513-5 <https://doi.org/10.1007/s00705-019-04513-5>`_
* Kochetkova *et al.* (2019) "*Tepidiforma bonchosmolovskayae* gen. nov., sp. nov., a moderately thermophilic *Chloroflexi* bacterium from a Chukotka hot spring (Arctic, Russia), representing a novel class, *Tepidiformia*, which includes the previously uncultivated lineage OLB14" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.003902 <https://doi.org/10.1099/ijsem.0.003902>`_
* Kovaleva *et al.* (2019) "*Tautonia sociabilis* gen. nov., sp. nov., a novel thermotolerant planctomycete, isolated from a 4000 m deep subterranean habitat" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.003467 <https://doi.org/10.1099/ijsem.0.003467>`_
* Labuda *et al.* (2019) "Bloodstream Infections With a Novel Nontuberculous Mycobacterium Involving 52 Outpatient Oncology Clinic Patients―Arkansas, 2018" *Clin. Inf. Dis.* `doi:10.1093/cid/ciz1120 <https://doi.org/10.1093/cid/ciz1120>`_
* Lan *et al.* (2019) "*Vogesella urethralis* *sp. nov.*, isolated from human urine, and emended descriptions of *Vogesella perlucida* and *Vogesella mureinivorans*" *Int. J. Syst. Evol. Microbiol.* `doi:10.1099/ijsem.0.003802 <https://doi.org/10.1099/ijsem.0.003802>`_
* Lawson *et al.* (2019) "Breast milk-derived human milk oligosaccharides promote *Bifidobacterium* interactions within a single ecosystem" *ISME J.* `doi:0.1038/s41396-019-0553-2 <https://doi.org/0.1038/s41396-019-0553-2>`_
* Ma *et al.* (2019) "First report of *Dickeya fangzhongdai* causing soft rot of onion in New York State" *Plant Dis.* `doi:10.1094/PDIS-09-19-1940-PDN <https://doi.org/10.1094/PDIS-09-19-1940-PDN>`_
* Matteo-Estrada *et al.* (2019) "Phylogenomics Reveals Clear Cases of Misclassification and Genus-Wide Phylogenetic Markers for *Acinetobacter*" *Genome Biol. Evol.* `doi:10.1093/gbe/evz178 <https://doi.org/10.1093/gbe/evz178>`_
* McIntyre *et al.* (2019) "Single-molecule sequencing detection of N6-methyladenine in microbial reference materials" *Nat. Comm.* `doi:10.1038/s41467-019-08289-9 <https://doi.org/s41467-019-08289-9>`_
* Nordmann *et al.* (2019) "Complete genome sequence of the virus isolate vB_BthM-Goe5 infecting *Bacillus thuringiensis*" *Arch. Virol.* `doi:10.1007/s00705-019-04187-z <https://10.1007/s00705-019-04187-z>`_
* Paim *et al.* (2019) "Evaluation of niche adaptation features by genome data mining approach of *Escherichia coli* urinary and gastrointestinal strains" *PeerJ Preprints* `doi:10.7287/peerj.preprints.27720v1 <https://doi.org/10.7287/peerj.preprints.27720v1>`_
* Park *et al* (2019) "Complete genome sequence of acetate-producing *Klebsiella pneumoniae* L5-2 isolated from infant feces" *3Biotech* `doi:10.1007/s13205-019-1578-y <https://doi.org/10.1007/s13205-019-1578-y>`_
* Pedron & van Gijsegem (2019) "Diversity in the Bacterial Genus *Dickeya* Grouping Plant Pathogens and Waterways Isolates" *OBM Genetics* `doi:10.21926/obm.genet.1904098 <https://doi.org/10.21926/obm.genet.1904098>`_
* Portier *et al.* (2019) "Elevation of *Pectobacterium carotovorum* subsp. *odoriferum* to species level as *Pectobacterium odoriferum* sp. nov., proposal of *Pectobacterium brasiliense* sp. nov. and *Pectobacterium actinidiae* sp. nov., emended description of *Pectobacterium carotovorum* and description of *Pectobacterium versatile* sp. nov., isolated from streams and symptoms on diverse plants" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.003611 <https://doi.org/10.1099/ijsem.0.003611>`_
* Potter *et al.* (2019) "In Silico Analysis of *Gardnerella* Genomospecies Detected in the Setting of Bacterial Vaginosis" *Clin. Chem.* `doi:10.1373/clinchem.2019.305474 <https://doi.org/10.1373/clinchem.2019.305474>`_
* Reichler *et al.* (2019) "A century of gray: A genomic locus found in 2 distinct *Pseudomonas* spp. is associated with historical and contemporary color defects in dairy products worldwide" *J. Dairy Sci.* `doi:10.3168/jds.2018-16192 <https://doi.org/10.3168/jds.2018-16192>`_
* Royo-Llonch *et al.* "Ecological and functional capabilities of an uncultured *Kordia* sp" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2019.126045 <https://doi.org/10.1016/j.syapm.2019.126045>`_
* Ruiz *et al.* (2019) "Microbiota of human precolostrum and its potential role as a source of bacteria to the infant mouth" *Sci. Rep.* `doi:10.1038/s41598-019-42514-1 <https://doi.org/10.1038/s41598-019-42514-1>`_
* Sant'Anna *et al.* (2019) "Genomic metrics made easy: what to do and where to go in the new era of bacterial taxonomy" *Crit. Rev. Microbiol.* `doi:10.1080/1040841X.2019.1569587 <https://doi.org/10.1080/1040841X.2019.1569587>`_
* Schmuhl *et al.* (2019) "Comparative Transcriptomic Profiling of *Yersinia enterocolitica* O:3 and O:8 Reveals Major Expression Differences of Fitness- and Virulence-Relevant Genes Indicating Ecological Separation" *mSystems* `doi:10.1128/mSystems.00239-18 <https://doi.org/10.1128/mSystems.00239-18>`_
* Spirina *et al.* (2019) "Draft Genome Sequence of Microbacterium sp. Gd 4-13, Isolated from Gydanskiy Peninsula Permafrost Sediments of Marine Origin" *Microb. Res. Announce.* `doi:10.1128/MRA.00889-19 <https://doi.org/10.1128/MRA.00889-19>`_
* Stefanic *et al.* (2019) "Intra-species DNA exchange: *Bacillus subtilis* prefers sex with less related strains" *BioRxiv* `doi:10.1101/756569 <https://doi.org/10.1101/756569>`_
* Stevens *et al.* (2019) "Whole-genome-based phylogeny of *Bacillus cytotoxicus* reveals different clades within the species and provides clues on ecology and evolution" *Sci. Rep.* `doi:10.1038/s41598-018-36254-x <https://doi.org/10.1038/s41598-018-36254-x>`_
* Tanaka *et al.* (2019) "Draft Genome Sequences of *Enterococcus faecalis* Strains Isolated from Healthy Japanese Individuals" *Microb. Res. Announce.* `doi:10.1128/MRA.00832-19 <https://doi.org/10.1128/MRA.00832-19>`_
* Thorell *et al.* (2019) "Isolates from colonic spirochaetosis in humans show high genomic divergence and carry potential pathogenic features but are not detected by 16S amplicon sequencing using standard primers for the human microbiota" *BioRxiv* `doi:doi.org/10.1101/544502 <https://doi.org/doi.org/10.1101/544502>`_
* Tian *et al.* (2019) "LINbase: A Web service for genome-based identification of microbes as members of crowdsourced taxa" *BioRxiv* `doi:10.1101/752212 <https://doi.org/10.1101/752212>`_
* Tohno *et al.* (2019) "*Lactobacillus salitolerans* sp. nov., a novel lactic acid bacterium isolated from spent mushroom substrates" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.003224 <https://doi.org/10.1099/ijsem.0.003224>`_
* Vazquez-Campos *et al.* (2019) "Genomic insights into the Archaea inhabiting an Australian radioactive legacy site" *BioRxiv* `doi:10.1101/728089 <https://doi.org/10.1101/728089>`_
* Vincent *et al.* (2019) "A Mesophilic *Aeromona salmonicida* Strain Isolated from an Unsuspected Host, the Micratory Bird Pied Avocet" *Microorganisms* `doi:10.3390/microorganisms7120592 <https://doi.org/10.3390/microorganisms7120592>`_
* Vincent *et al.* (2019) "Investigation of the virulence and genomics of *Aeromonas salmonicida* strains isolated from human patients" *Inf. Genet. Evol.* `doi:10.1016/j.meegid.2018.11.019 <https://10.1016/j.meegid.2018.11.019>`_
* Vincent *et al.* (2019) "Revisiting the taxonomy and evolution of pathogenicity of the genus *Leptospira* through the prism of genomics" *PLoS Neg. Trop. Dis.* `doi:10.1371/journal.pntd.0007270 <https://doi.org/10.1371/journal.pntd.0007270>`_
* Wallner *et al.* (2019) "Genomic analyses of *Burkholderia cenocepacia* reveal multiple species with differential host-adaptation to plants and humans" *BMC Genomics* `doi:10.1186/s12864-019-6186-z <https://doi.org/10.1186/s12864-019-6186-z>`_
* Wang *et al.* (2019) "Occurrence of CTX-M-123-producing *Salmonella* Indiana in chicken carcasses: a new challenge for the poultry industry and food safety" *J. Antimicrob. Chemo.* `doi:10.1093/jac/dkz386 <https://doi.org/10.1093/jac/dkz386>`_
* Webster *et al.* (2019) "Genome Sequences of Two Choline-Utilizing Methanogenic Archaea, *Methanococcoides* spp., Isolated from Marine Sediments" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00342-19 <https://dx.doi.org/10.1128/MRA.00342-19>`_
* Webster *et al.* (2019) "The Genome Sequences of Three *Paraburkholderia* sp. Strains Isolated from Wood-Decay Fungi Reveal Them as Novel Species with Antimicrobial Biosynthetic Potential" *Microbiol. Res. Ann.* `doi:10.1128/MRA.00778-19 <https://dx.doi.org/10.1128/MRA.00778-19>`_
* Wiegand *et al.* (2019) "Cultivation and functional characterization of 79 planctomycetes uncovers their unique biology" *Nat. Microbiol.* `doi:10.1038/s41564-019-0588-1 <https://doi.org/10.1038/s41564-019-0588-1>`_
* Wittouck *et al.* (2019) " A genome-based species taxonomy of the *Lactobacillus* genus complex" *mSystems* `doi:10.1128/mSystems.00264-19 <https://doi.org/10.1128/mSystems.00264-19>`_
* Yin *et al.* (2019) "A hybrid sub-lineage of *Listeria monocytogenes* comprising hypervirulent isolates" *Nat. Comm.* `doi:10.1038/s41467-019-12072-1 <https://doi.org/10.1038/s41467-019-12072-1>`_
* Yin *et al.* (2019) "Genetic Diversity of *Listeria monocytogenes* Isolates from Invasive Listeriosis in China" *Foodborne Path. Dis.* `doi:10.1089/fpd.2019.2693 <https://doi.org/10.1089/fpd.2019.2693>`_
* Zabel *et al.* (2019) "Novel Genes and Metabolite Trends in *Bifidobacterium longum* subsp. *infantis* Bi-26 Metabolism of Human Milk Oligosaccharide 2′-fucosyllactose" *Sci. Rep.* `doi:10.1038/s41598-019-43780-9 <https://doi.org/s41598-019-43780-9>`_
* Zakham *et al.* (2019) "Molecular diagnosis and enrichment culture identified a septic pseudoarthrosis due to an infection with *Erysipelatoclostridium ramosum*" *Int. J. Inf. Dis.* `doi:10.1016/j.ijid.2019.02.001 <https://doi.org/10.1016/j.ijid.2019.02.001>`_
* Zhu *et al.* (2019) "First Report of Integrative Conjugative Elements in *Riemerella anatipestifer* Isolates From Ducks in China" *Front. Vet. Sci.* `doi:10.3389/fvets.2019.00128 <https://doi.org/10.3389/fvets.2019.00128>`_
* Zhu *et al.* (2019) "Pan-genome analysis of *Riemerella anatipestifer* reveals its genomic diversity and acquired antibiotic resistance associated with genomic islands" *Func. Int. Genom* `doi:10.1007/s10142-019-00715-x <https://doi.org/10.1007/s10142-019-00715-x>`_

^^^^
2018
^^^^

* Alex & Antunes (2018) "Genus-wide comparison of *Pseudovibrio* bacterial genomes reveal diverse adaptations to different marine invertebrate hosts" *PLoS One* `doi:10.1371/journal.pone.0194368 <https://doi.org/10.1371/journal.pone.0194368>`_
* Beaton *et al.* (2018) "Community-led comparative genomic and phenotypic analysis of the aquaculture pathogen *Pseudomonas baetica* a390T sequenced by Ion semiconductor and Nanopore technologies" *FEMS Micro. Lett.* `doi:10.1093/femsle/fny069 <https://doi.org/10.1093/femsle/fny069>`_
* Bogema *et al.* (2018) "Analysis of *Theileria orientalis* draft genome sequences reveals potential species-level divergence of the Ikeda, Chitose and Buffeli genotypes" *BMC Genomics* `doi:10.1186/s12864-018-4701-2 <https://doi.org/10.1186/s12864-018-4701-2>`_
* Brand *et al.* (2018) "Niche Differentiation among Three Closely Related *Competibacteraceae* Clades at a Full-Scale Activated Sludge Wastewater Treatment Plant and Putative Linkages to Process Performance" *App. Env. Micro.* `doi:10.1128/AEM.02301-18 <https://doi.org/10.1128/AEM.02301-18>`_
* Bridel *et al.* (2018) "Comparative Genomics of *Tenacibaculum dicentrarchi* and “*Tenacibaculum finnmarkense*” Highlights Intricate Evolution of Fish-Pathogenic Species" *Genome Biol. Evol.* `doi:10.1093/gbe/evy020 <https://doi.org/10.1093/gbe/evy020>`_
* Carlos *et al.* (2018) "Substrate Shift Reveals Roles for Members of Bacterial Consortia in Degradation of Plant Cell Wall Polymers" *Front. Microbiol.* `doi:10.3389/fmicb.2018.00364 <https://doi.org/10.3389/fmicb.2018.00364>`_
* Covarrubias *et al.* (2018) "Occurrence, integrity and functionality of *Aca*ML1–like viruses infecting extreme acidophiles of the *Acidithiobacillus* species complex" *Res. Microbiol.* `doi:10.1016/j.resmic.2018.07.005 <http://doi.org/10.1016/j.resmic.2018.07.005>`_
* da Gama *et al.* (2018) "Taxonomic Repositioning of *Xanthomonas campestris* pv. *viticola* (Nayudu 1972) Dye 1978 as *Xanthomonas citri* pv. *viticola* (Nayudu 1972) Dye 1978 comb. nov. and Emendation of the Description of *Xanthomonas citri* pv. *anacardii* to Include Pigmented Isolates Pathogenic to Cashew Plant" *Phytopath.* `doi:10.1094/PHYTO-02-18-0037-R <https://doi.org/10.1094/PHYTO-02-18-0037-R>`_
* Ferretti *et al.* (2018) "Mother-to-Infant Microbial Transmission from Different Body Sites Shapes the Developing Infant Gut Microbiome" *Cell Host Microbe* `doi:10.1016/j.chom.2018.06.005 <https://doi.org/10.1016/j.chom.2018.06.005>`_
* Fontana *et al.* (2018) "Genetic Signatures of Dairy *Lactobacillus casei* Group" *Front. Microbiol.* `doi:10.3389/fmicb.2018.02611 <https://doi.org/10.3389/fmicb.2018.02611>`_
* Freschi *et al.* (2018) "The *Pseudomonas aeruginosa* Pan-Genome Provides New Insights on Its Population Structure, Horizontal Gene Transfer, and Pathogenicity" *Genome Biol. Evol.* `doi:10.1093/gbe/evy259 <https://doi.org/10.1093/gbe/evy259>`_
* Gillis *et al.* (2018) "Role of plasmid plasticity and mobile genetic elements in the entomopathogen *Bacillus thuringiensis* serovar *israelensis*" *FEMS Micro. Rev.* `doi:10.1093/femsre/fuy034 <https://doi.org/10.1093/femsre/fuy034>`_
* Gragna-Miraglia *et al.* (2018) "Phylogenomics picks out the par excellence markers for species phylogeny in the genus *Staphylococcus*" *PeerJ* `doi:10.7717/peerj.5839 <https://doi.org/10.7717/peerj.5839>`_
* Hubbard *et al.* (2018) "Comparison of the first whole genome sequence of ‘*Haemophilus quentini*’ with two new strains of ‘*Haemophilus quentini*’ and other species of *Haemophilus*" *Genome* `doi:10.1139/gen-2017-0195 <https://doi.org/10.1139/gen-2017-0195>`_
* Issotta *et al.* (2018) "Insights into the biology of acidophilic members of the *Acidiferrobacteraceae* family derived from comparative genomic analyses" *Res. Microbiol.* `doi:10.1016/j.resmic.2018.08.001 <https://doi.org/10.1016/j.resmic.2018.08.001>`_
* Jangam *et al.* (2018) "Draft Genome Sequence of *Vibrio parahaemolyticus* Strain VP14, Isolated from a *Penaeus vannamei* Culture Farm" *Micro. Res. Ann.* `doi:10.1128/genomeA.00149-18 <https://10.1128/genomeA.00149-18>`_
* Jarett *et al.* (2018) "Single-cell genomics of co-sorted *Nanoarchaeota* suggests novel putative host associations and diversification of proteins involved in symbiosis" *Microbiome* `doi:10.1186/s40168-018-0539-8 <https://doi.org/10.1186/s40168-018-0539-8>`_
* Jung *et al.* (2018) "Complete genome sequence of *Bifidobacterium choerinum* FMB-1, a resistant starch-degrading bacterium" *J. Biotech.* `doi:10.1016/j.jbiotec.2018.03.009 <https://doi.org/10.1016/j.jbiotec.2018.03.009>`_
* Lazarte *et al.* (2018) "*Bacillus wiedmannii* biovar *thuringiensis*: A Specialized Mosquitocidal Pathogen with Plasmids from Diverse Origins" *Genome Biol. Evol.* `doi:10.1093/gbe/evy211 <https://doi.org/10.1093/gbe/evy211>`_
* Li *et al.* (2018) "A Novel Strategy for Detecting Recent Horizontal Gene Transfer and Its Application to *Rhizobium* Strains" *Front. Microbiol.* `doi:10.3389/fmicb.2018.00973 <https://dx.doi.org/10.3389/fmicb.2018.00973>`_
* Lima *et al.* "Genome sequencing and functional characterization of the non-pathogenic *Klebsiella pneumoniae* KpGe bacteria* *Microbes Inf.* `doi:10.1016/j.micinf.2018.04.001 <https://doi.org/10.1016/j.micinf.2018.04.001>`_
* McCann *et al.* (2018) "Viromes of one year old infants reveal the impact of birth mode on microbiome diversity" *PeerJ* `doi:10.7717/peerj.4694 <https://doi.org/10.7717/peerj.4694>`_
* Morales-Covarrubias (2018) "*Streptococcus penaeicida* sp. nov., isolated from a diseased farmed Pacific white shrimp (*Penaeus vannamei*)" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.002693 <https://doi.org/10.1099/ijsem.0.002693>`_
* Munoz-Villagran *et al.* (2018) "Comparative genomic analysis of a new tellurite-resistant *Psychrobacter* strain isolated from the Antarctic Peninsula" *PeerJ* `doi:10.7717/peerj.4402 <https://doi.org/10.7717/peerj.4402>`_
* Nascimento *et al.* (2018) "From plants to nematodes: *Serratia grimesii* BXF1 genome reveals an adaptation to the modulation of multi-species interactions" *Microb. Genom.* `doi:10.1099/mgen.0.000178 <https://doi.org/10.1099/mgen.0.000178>`_
* Orr *et al.* (2018) "De novo assembly of the *Pasteuria penetrans* genome reveals high plasticity, host dependency, and BclA-like collagens" *BioRxiv* `doi:10.1101/485748 <https://doi.org/10.1101/485748>`_
* Pinto *et al.* (2018) "Draft Genome Sequences of Novel *Pseudomonas*, *Flavobacterium*, and *Sediminibacterium* Strains from a Freshwater Ecosystem" *Micro. Res. Ann.* `doi:10.1128/genomeA.00009-18 <https://doi.org/10.1128/genomeA.00009-18>`_
* Potter *et al.* (2018) "Population Structure, Antibiotic Resistance, and Uropathogenicity of *Klebsiella variicola*" *mBio* `doi:10.1128/mBio.02481-18 <https://doi.org/10.1128/mBio.02481-18>`_
* Potter *et al.* (2018) "*Superficieibacter electus* gen. nov., sp. nov., an Extended-Spectrum β-Lactamase Possessing Member of the Enterobacteriaceae Family, Isolated From Intensive Care Unit Surfaces" *Front. Microbiol.* `doi:10.3389/fmicb.2018.01629 <https://doi.org/10.3389/fmicb.2018.01629>`_
* Samad *et al.* (2017) "Comparative genome analysis of the vineyard weed endophyte *Pseudomonas viridiflava* CDRTc14 showing selective herbicidal activity" *Sci. Rep.* `doi:10.1038/s41598-017-16495-y <https://doi.org/10.1038/s41598-017-16495-y>`_
* Sant'Anna *et al.* (2018) "Genome-based reclassification of *Paenibacillus dauci* as a later heterotypic synonym of *Paenibacillus shenyangensis*" *Int. J. Syst. Evol. Micro.* `doi:10.1099/ijsem.0.003127 <https://10.1099/ijsem.0.003127>`_
* Schilling *et al.* (2018) "Genomic Analysis of the Recent Viral Isolate vB_BthP-Goe4 Reveals Increased Diversity of φ29-Like Phages" *Viruses* `doi:10.3390/v10110624 <https://doi.org/10.3390/v10110624>`_
* Stevens *et al.* (2018) "Massive Diversity in Whole-Genome Sequences of *Streptococcus suis* Strains from Infected Pigs in Switzerland" *Microbiol. Res. Ann.* `doi:10.1128/MRA.01656-18 <https://dx.doi.org/10.1128/MRA.01656-18>`_
* Tanizawa *et al.* (2018) "Lactobacillus paragasseri sp. nov., a sister taxon of Lactobacillus gasseri, based on whole-genome sequence analyses" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.003020 <https://doi.org/10.1099/ijsem.0.003020>`_
* Vincent & Charette (2018) "Completion of genome of *Aeromonas salmonicida* subsp. *salmonicida* 01-B526 reveals how sequencing technologies can influence sequence quality and result interpretations" *New Microb. New Inf.* `doi:10.1016/j.nmni.2018.05.007 <https://doi.org/10.1016/j.nmni.2018.05.007>`_
* Wilhelm (2018) "Following the terrestrial tracks of *Caulobacter* - redefining the ecology of a reputed aquatic oligotroph" *ISME J* `doi:10.1038/s41396-018-0257-z <https://doi.org/10.1038/s41396-018-0257-z>`_
* Wittwer *et al.* (2018) "Population Genomics of *Francisella tularensis* subsp. *holarctica* and its Implication on the Eco-Epidemiology of Tularemia in Switzerland" *Front. Cell. Inf. Microbiol.* `doi:10.3389/fcimb.2018.00089 <https://doi.org/10.3389/fcimb.2018.00089>`_
* Zhang *et al.* (2018) "Draft Genome Sequence of *Komagataeibacter maltaceti* LMG 1529T, a Vinegar-Producing Acetic Acid Bacterium Isolated from Malt Vinegar Brewery Acetifiers" *Micro. Res. Ann.* `doi:10.1128/genomeA.00330-18 <https://doi.org/10.1128/genomeA.00330-18>`_

^^^^
2017
^^^^

* Anderson *et al.* (2017) "Genomic variation in microbial populations inhabiting the marine subseafloor at deep-sea hydrothermal vents" *Nat. Comm.* `doi:10.1038/s41467-017-01228-6 <https://doi.org/10.1038/s41467-017-01228-6>`_
* Ding *et al.* (2017) "Loss of the ssrA genome island led to partial debromination in the PBDE respiring *Dehalococcoides mccartyi* strain GY50" *Env. Micro.* `doi:10.1111/1462-2920.13817 <https://doi.org/10.1111/1462-2920.13817>`_
* Edgington *et al.* (2017) "Genome Sequences of Chancellor, Mitti, and Wintermute, Three Subcluster K4 Phages Isolated Using *Mycobacterium smegmatis* mc^{2}155" *Microbiol. Res. Ann.* `doi:10.1128/genomeA.01070-17 <https://doi.org/10.1128/genomeA.01070-17>`_
* Esposito *et al.* (2017) "Evolution of *Stenotrophomonas maltophilia* in Cystic Fibrosis Lung over Chronic Infection: A Genomic and Phenotypic Population Study" *Front. Microbiol.* `doi:10.3389/fmicb.2017.01590 <https://10.3389/fmicb.2017.01590>`_
* Jeukens *et al.* (2017) "A Pan-Genomic Approach to Understand the Basis of Host Adaptation in *Achromobacter*" *Genome Biol. Evol.* `doi:10.1093/gbe/evx061 <https://doi.org/10.1093/gbe/evx061>`_
* Ke *et al.* (2017) "Comparative genomics of *Vibrio campbellii* strains and core species of the *Vibrio Harveyi* clade" *Sci. Rep.* `doi:10.1038/srep41394 <https://doi.org/10.1038/srep41394>`_
* Kumar *et al.* (2017) "Draft Genome Sequence of the Luminescent Strain *Vibrio campbellii* LB102, Isolated from a Black Tiger Shrimp (*Penaeus monodon*) Broodstock Rearing System" *Micro. Res. Ann.* `doi:10.1128/genomeA.00342-17 <https://doi.org/10.1128/genomeA.00342-17>`_
* Pelve *et al.* (2017) "Bacterial Succession on Sinking Particles in the Ocean's Interior" *Front. Microbiol.* `doi:10.3389/fmicb.2017.02269 <https://doi.org/10.3389/fmicb.2017.02269>`_
* Poehlein *et al.* (2017) "Microbial solvent formation revisited by comparative genome analysis" *Biotech. Biofuels* `doi:10.1186/s13068-017-0742-z <https://doi.org/10.1186/s13068-017-0742-z>`_
* Ruiz-Valdeviezo *et al.* (2017) "Complete Genome Sequence of a Novel Nonnodulating *Rhizobium* Species Isolated from *Agave americana* L. Rhizosphere" *Micro. Res. Ann.* `doi:10.1128/genomeA.01280-17 <https://doi.org/10.1128/genomeA.01280-17>`_
* Tada *et al.* (2017) "Revealing the genomic differences between two subgroups in *Lactobacillus gasseri*" *Biosci. Microb. Food Health* `doi:10.12938/bmfh.17-006 <https://doi.org/10.12938/bmfh.17-006>`_
* Tanizawa *et al.* (2017) "Genomic characterization reconfirms the taxonomic status of *Lactobacillus parakefiri*" *Biosci. Microb. Food Health* `doi:10.12938/bmfh.16-026 <https://doi.org/10.12938/bmfh.16-026>`_
* Tohno *et al.* (2017) "*Lactobacillus silagincola* sp. nov. and *Lactobacillus pentosiphilus* sp. nov., isolated from silage" *Int. J Syst. Evol. Biol* `doi:10.1099/ijsem.0.002196 <https://doi.org/10.1099/ijsem.0.002196>`_
* Vincent *et al.* (2017) "Study of mesophilic *Aeromonas salmonicida* A527 strain sheds light on the species’ lifestyles and taxonomic dilemma" *FEMS Micro. Lett.* `doi:10.1093/femsle/fnx239 <https://doi.org/10.1093/femsle/fnx239>`_
* Vollmers *et al.* (2017) "Untangling Genomes of Novel *Planctomycetal* and *Verrucomicrobial* Species from Monterey Bay Kelp Forest Metagenomes by Refined Binning" *Front. Microbiol.* `doi:10.3389/fmicb.2017.00472 <https://doi.org/10.3389/fmicb.2017.00472>`_
* Wang *et al.* (2017) "Genomic sequence of 'Candidatus *Liberibacter solanacearum*' haplotype C and its comparison with haplotype A and B genomes" *PLoS One* `doi:10.1371/journal.pone.0171531 <https://doi.org/10.1371/journal.pone.0171531>`_

^^^^
2016
^^^^

* Burstein *et al.* (2016) "New CRISPR–Cas systems from uncultivated microbes" *Nature* `doi:10.1038/nature21059 <https://doi.org/10.1038/nature21059>`_
* Gupta *et al.* (2016) "Comparative genomic analysis of novel *Acinetobacter* symbionts: A combined systems biology and genomics approach" *Sci. Rep.* `doi:10.1038/srep29043 <https://doi.org/srep29043>`_
* Haack *et al.* (2016) "Molecular Keys to the *Janthinobacterium* and *Duganella* spp. Interaction with the Plant Pathogen *Fusarium graminearum*" *Front. Microbiol.* `doi:10.3389/fmicb.2016.01668 <https://dx.doi.org/10.3389/fmicb.2016.01668>`_
* Maeno *et al.* (2016) "Genomic characterization of a fructophilic bee symbiont *Lactobacillus kunkeei* reveals its niche-specific adaptation" *Syst. Appl. Microbiol.* `doi:10.1016/j.syapm.2016.09.006 <https://doi.org/10.1016/j.syapm.2016.09.006>`_
* Pritchard *et al.* (2016) "Genomics and taxonomy in diagnostics for food security: soft-rotting enterobacterial plant pathogens" *Anal. Methods* `doi:10.1039/C5AY02550H <https://doi.org/10.1039/C5AY02550H>`_
* Rodriguez-Rojas *et al.* (2016) "Draft Genome Sequence of a Multi-Metal Resistant Bacterium *Pseudomonas putida* ATH-43 Isolated from Greenwich Island, Antarctica" *Front. Microbiol.* `doi:10.3389/fmicb.2016.01777 <https://doi.org/10.3389/fmicb.2016.01777>`_
* Tanizawa *et al.* (2016) "DFAST and DAGA: web-based integrated genome annotation tools and resources" *Biosci. Microb. Food Health* `doi:10.12938/bmfh.16-003 <https://doi.org/10.12938/bmfh.16-003>`_
* Zheng *et al.* (2016) "Metabolism of Toxic Sugars by Strains of the Bee Gut Symbiont *Gilliamella apicola*" *mBio* `doi:10.1128/mBio.01326-16 <https://doi.org/10.1128/mBio.01326-16>`_


.. _10.1039/C5AY02550H: https://dx.doi.org/10.1039/C5AY02550H
